File: chapter4/the_sum_of_a_range.rb
def range(start, last, step = start < last ? 1 : -1)
  if start > last
    start, last = last, start
    start.step(last, step).to_a.reverse
  else
    start.step(last, step).to_a
  end
end

def sum(array)
  total = 0
  array.each do |n|
    total += n
  end
  total
end

## 畳み込み演算
def sum(array)
  array.inject(:+)
end

p range(10, 3, -1)
p range(2, 10, 3)
p sum(range(1, 10))

File: chapter4/the_sum_of_a_range.js
const range = (start, end) => [...new Array(end - start + 1).keys()].map(n => n + start)

const sum = arr => arr.reduce((acc, value) => acc + value)

console.log(range(1, 10))
console.log(sum(range(1, 10)))

// モブプロしたやつ
{
  const range = (start, end, step = start < end ? 1 : -1) => {
    let result = [];
    if (step > 0) {
      for (let i = start; i <= end; i += step) {
        result.push(i);
      }
    } else if (step < 0) {
      for(let i = start;i >= end; i += step) {
        result.push(i)
      }
    }
    return result
  }
}

{
  const range = (start, end, step = start < end ? 1 : -1) => {
    if (step === 0 || (end - start) * step < 0) return []
    return [...new Array(Math.ceil(((Math.abs(end - start) + 1) / Math.abs(step)))).keys()].map(n => n * step + start)
  }
}

// 上のやつを読みやすくしたやつ

{
  const range = (start, end, step = start < end ? 1 : -1) => {
    if (step === 0 || (end - start) * step < 0) return [];
    const size = Math.ceil((Math.abs(end - start) + 1) / Math.abs(step));
    return [...new Array(size).keys()].map(n => n * step + start);
  }
}
